### PR TITLE
Respect reduction of inline_for_extraction qualifiers even under norm requests during extraction

### PR DESCRIPTION
**Pull request by nikswamy**
Consider this example:

```
module Extract
assume
val something_that_blocks : bool
inline_for_extraction
let f (g:nat) =
  if something_that_blocks then g else 0
let rec sum_to (n:nat) : nat =
  if n = 0 then 0 else n + sum_to (n - 1)
let test1 = f (sum_to 17)
let test2 = norm [zeta; delta_only [`%sum_to]; iota; primops] (f (sum_to 17))
let test3 = norm [zeta_full; delta_only [`%sum_to]; iota; primops] (f (sum_to 17))
```

If you extract it to OCaml, prior to this PR, you would get:

```
open Prims
let (something_that_blocks : Prims.bool) =
  Obj.magic
    (fun uu___ -> failwith "Not yet implemented:something_that_blocks")
let (f : Prims.nat -> Prims.int) =
  fun g -> if something_that_blocks then g else Prims.int_zero
let rec (sum_to : Prims.nat -> Prims.nat) =
  fun n ->
    if n = Prims.int_zero
    then Prims.int_zero
    else n + (sum_to (n - Prims.int_one))
let (test1 : Prims.int) =
  if something_that_blocks
  then sum_to (Prims.of_int (17))
  else Prims.int_zero
let (test2 : Prims.int) = f (Prims.of_int (153))
let (test3 : Prims.int) = f (Prims.of_int (153))
```

`test2` and `test3` are surprising, i.e., the inline on `f` was not honored. That's because there was a norm request on `test2` and `test3` which masked the inlining requests that happen by default in extraction. That's what this PR is trying to fix.

Now, with this PR you get:

```
let (test1 : Prims.int) =
  if something_that_blocks
  then sum_to (Prims.of_int (17))
  else Prims.int_zero
let (test2 : Prims.int) =
  if something_that_blocks
  then sum_to (Prims.of_int (17))
  else Prims.int_zero
let (test3 : Prims.int) =
  if something_that_blocks then (Prims.of_int (153)) else Prims.int_zero
```

`test1` is unchanged

`test2` is perhaps surprising: what has happened here is that we unfolded `f` first, since it is `inline_for_extraction`. But, that produced an unreducible match and the recursive function `sum_to 17` is left unreduced under the blocked match

`test3` gives us what we probably want, i.e., use zeta_full to not disable zeta under blocked matches

So, as the title says, this PR makes the normalizer respect `inline_for_extraction` and `unfold` in explicit user-provided `norm` calls during extraction. But, this change in behavior can also lead to surprises. See, e.g., https://github.com/project-everest/everparse/commit/d42a3cefc323dd0d7a2eb7fc6a06dbb959845a79

### CODE REVIEW

**Comment by tahina-pro** [comment]
I should complete the following tasks before this PR can be merged:
* [ ] merge project-everest/everparse@d42a3cefc323dd0d7a2eb7fc6a06dbb959845a79
* [ ] merge project-everest/everquic-crypto@4d9710e65fbcece695a83c131e98741adb54962e

File: src/typechecker/FStar.TypeChecker.Normalize.fs [hunk 1]
```diff
@@ -942,39 +965,37 @@ let should_unfold cfg should_reify fv qninfo : should_unfold_res =
...
-            BU.for_some
-              (fun q ->
-                BU.for_some 
-                  (fun qual -> Print.qual_to_string qual = q)
-                  quals)
-              qs)
-        ]
+        let meets_some_criterion =
+            comb_or [
+            (if cfg.steps.for_extraction
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Normalize.fs:970]: This is the main change: If we're in extraction, then always try to unfold `inline_for_extraction` and `unfold` qualified definitions

File: src/typechecker/FStar.TypeChecker.Normalize.fs [hunk 2]
```diff
@@ -81,8 +95,7 @@ type stack_elt =
  | CBVApp   of env * term * aqual * Range.range
  | Meta     of env * S.metadata * Range.range
  | Let      of env * binders * letbinding * Range.range
- | Cfg      of cfg
- | Debug    of term * BU.time
+ | Cfg      of cfg * option<(term * BU.time)>
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Normalize.fs:98]: This change fuses the Debug and Cfg nodes, by adding an optional debug value in a Cfg. 

The motivation for this is that prior to this change, if you turned on debugging using --debug_level print_normalized_terms, then you would actually change the behavior of the normalizer ... which is crazy.

The reason for this behavior was that adding a Debug marker on the stack causes many invariants to be broken. Fusing the debug marker with the Cfg marker ensures that debugging is handled wherever Cfg changes take place, allowing debug info to be emitted without changing the behavior.

Moral of the story: Don't add another kind of element on the stack, unless you're really sure that you are really consuming that element off the stack everywhere you need to.

File: src/typechecker/FStar.TypeChecker.Normalize.fs [hunk 3]
```diff
@@ -1186,15 +1207,23 @@ let rec norm : cfg -> env -> stack -> term -> term =
               let delta_level =
                 if s |> BU.for_some (function UnfoldUntil _ | UnfoldOnly _ | UnfoldFully _ -> true | _ -> false)
                 then [Unfold delta_constant]
-                else [NoDelta] in
-              let cfg' = {cfg with steps = ({ to_fsteps s with in_full_norm_request=true})
+                else if cfg.steps.for_extraction
+                then [Env.Eager_unfolding_only; Env.InliningDelta]
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Normalize.fs:1211]: If no other delta steps are set, and we're in extraction, then at least set it to unfolding inline_for_extraction and unfold definitions

**Pull request merged**